Enterprise-CMCS/managed-care-review
# MCR-3879: Refine matching rate revision to contract revision

## Summary
[MCR-3879](https://qmacbis.atlassian.net/browse/MCR-3879)

This change is to allow `contractWithHistoryToDomainModel` to still find the correct rate revision with the [Edit & Unlock Rate](https://qmacbis.atlassian.net/browse/MCR-3777) feature work. 

- Update `contractWithHistoryToDomainModel` to find the rate revision for the contract revision when rates can be independently submitted and unlocked.
- Update `rateWithHistoryToDomainModel` to find the contract revision for the rate revision when rates can be independently submitted and unlocked.
- I decided not to pull out the part of `contractWithHistoryToDomainModel` that finds the rate revision into its own function.
   - Unit testing this function with mock data would have been difficult. Since it is embedded into the `findContractWithHistory` and `findRateWithHistory` hanlders, tests for matching revisions are done there.
- Updated docs. 
- Fix `submitRate` and `unlockRate` so `findFirst` starts in descending order.

#### Related issues

#### Screenshots

#### Test cases covered

`findContractWithHistory.test.ts`
- `'matches correct rate revisions to contract revision with independent rate unlocks and submits'`

`findRateWithHistory.test.ts`
- `'matches contract revision to rate revision with independent rate submit and unlocks'`



## QA guidance

Some existing API tests do test for the correct rate revision, and they all passed, so I think the change will result will be the same, just expanded to accommodate independent rate unlock/resubmit. Just to make sure, I would manually unlock and resubmit submissions with rate changes and make sure latest submission and previous submissions display the correct rate data.
